### PR TITLE
VTOL_TAKEOFF, VTOL_LAND commands and VT_FORCE_VTOL param

### LINKED ISSUE #3492: VTOL: Support MAVLink commands for direct VTOL takeoff/land
https://github.com/mavlink/mavlink/pull/487

### LINKED ISSUE #3712: VTOL: implement a parameter enforcing auto takeoff and land are always in vtol mode
Most VTOL airframes (like tailsitters and some tiltrotors and quadplanes) are not capable of fixed wing takeoff or landing because they lack proper landing gear. A LAND command (from mission, failsafe, switch or offboard) would result in undesired behavior.

A parameter like VT_ENFORCE_VTOL should be added to indicate the vehicle will never attempt to takeoff or land in fixed wing mode unless there is no other choice (failed back transition, out of power).

### PR DESCRIPTION
**Pull request by sanderux**
support for MAV_CMD_NAV_VTOL_TAKEOFF and MAV_CMD_NAV_VTOL_LAND 
Also implement VT_FORCE_VTOL to force land as VTOL
replaces https://github.com/PX4/Firmware/pull/3720

Fixes #3492
Fixes #3712

### CODE REVIEW

**Comment by sanderux** [comment]
@AndreasAntener I cleaned up the mission code, sanitized the functionality, improved back transition behavior, fixed alignment and implemented force_vtol procedure.
SITL tested all reasonable scenarios and regression tested on the old procedure (takeoff -> trans -> wp -> trans -> land).

I'm especially happy with the back transition improvement, we could see if the basis could change from time to velocity but for now this seems to be a big improvement.

I will test fly this tree soon but i think it is ready to rebase and merge as i see no possible regression and it only implements previously unsupported functionality.

Somehow this tagged along through a submodule init: https://github.com/PX4/Firmware/pull/4055/files#diff-d62dbab64e619e0e5926b64585204efdR1
I think that should be resolved by rebasing but i'm not sure.

File: src/modules/vtol_att_control/vtol_att_control_params.c [hunk 1]
```diff
@@ -312,3 +312,12 @@ PARAM_DEFINE_FLOAT(VT_TRANS_TIMEOUT, 15.0f);
...
 PARAM_DEFINE_FLOAT(VT_TRANS_MIN_TM, 2.0f);
+
+/**
+ * Force VTOL mode takeoff and land
+ *
+ * @min 0
+ * @max 1
+ * @group VTOL Attitude Control
+ */
+PARAM_DEFINE_INT32(VT_FORCE_VTOL, 0);
```
Comment (LorenzMeier) [inline src/modules/vtol_att_control/vtol_att_control_params.c]: This should be defined in the navigator params, (NAV_FORCE_VTOL), as its only used there.

Comment (sanderux) [inline src/modules/vtol_att_control/vtol_att_control_params.c]: All vtol related params start with VT_ perhaps VT_NAV_FORCE_VTOL ?

**Comment by AndreasAntener** [comment]
@sanderux Could you rebase and squash the development commits? There are also wrong indentations in the mission code it seems

**Comment by LorenzMeier** [comment]
@AndreasAntener @sanderux There is a 15 character limit on parameter length, what I proposed works, what you proposed doesn't.

**Comment by sanderux** [comment]
@LorenzMeier I tested VT_NAV_FORCE_VTL as param and that works. Do you want me to change it?

**Comment by LorenzMeier** [comment]
16 is the hard limit without NUL-termination. It should work, but you can get yourself into corner cases in downstream implementations. I suggest to use VT_NAV_FORCE_VT instead.

**Comment by sanderux** [comment]
ready to merge

**Comment by LorenzMeier** [comment]
Is this state SITL and outdoor tested? I think it should be, as the changes are significant.

**Comment by sanderux** [comment]
Sitl is tested, outdoor tomorrow

**Comment by sanderux** [comment]
Outdoor tested in quite strong wind. It seemed to do some weathervaning but worked well.
http://logs.uaventure.com/view/JRG4pRghKoV3RK8eS3o75k#AS_Yaw_PLOT

During lading it weathervaned and got lift from it's wings, it therefor acted up a little.

**Comment by sanderux** [comment]
Here is a second log:
http://logs.uaventure.com/view/PJYu2ARidpuQyFsX9qxdKm

The yaw setpoint issue is consistent
The airspeed sensor for the quadranger acted up here (turned out to be the radio telem being too close)

**Comment by sanderux** [comment]
This should fix the yaw issue during transition

**Comment by AndreasAntener** [comment]
@sanderux before I forget, mission item reached: one reason to not wait on the transition to be finished is that the position controller won't know the next waypoint and therefore cannot control the position during the transition in case we later implement this. haven't thought about it in detail though.

**Comment by sanderux** [comment]
the advantage of this command is that the user can point the vehicle into the wind. It is basically the autonomous equivlant of flicking the aux1 switch.

For position control it should hold altitude and heading until ttansition has completed or timeout occurs.

**Comment by AndreasAntener** [comment]
Sure but it's changing the behaviour in general, also for the separate transitions (not part of VTOL tko/lnd). What we talked about yesterday, if we want the FW position controller to guide it through blending, it needs to know the next waypoint. If it doesn't get the actual next mission waypoint, we need to generate a temporary waypoint that lies in the direction of the target transition heading. Another option would be to look at the "next" waypoint in the triplet, which contains the next position during the transition. Ok, I think we have 2 good options here so it's fine.

Note: simply holding heading and attitude is not going to work, it will drift off path.

**Comment by sanderux** [comment]
Yes your right, i think a temp waypoint would be a good idea.

**Comment by LorenzMeier** [comment]
Waiting for conclusive test flights to merge this.

**Comment by AndreasAntener** [comment]
I did a sitl test to check on the Z velocity issue that i see in your log from #4079:

screen shot 2016-03-27 at 12 15 55

The position controller still sees a takeoff type setpoint during the transition. Needs to be reset, should be a small fix.

**Comment by sanderux** [comment]
@AndreasAntener do you agree on the location of setting the setpoint type?

**Comment by sanderux** [comment]
@AndreasAntener i added the virtual waypoint in front of the takeoff item. This should work fine now with your transition position holding

**Comment by sanderux** [comment]
Here is a log for a SITL flight based on this branch combined with https://github.com/PX4/Firmware/pull/4093. looks very smooth and functions as expected.
http://logs.uaventure.com/view/VXJnLmimbG8px8aiuPtBja

image

**Comment by sanderux** [comment]
Outdoor tested, looking good:
http://logs.uaventure.com/view/jg2rHZDRmAboYLiqyFVuge

**Comment by LorenzMeier** [comment]
Rebased and applied, thanks!

**Pull request closed without merging**